Application closes if MQTT server is unavailable

Observed today following a MQTT server upgrade.
To be investigated and documented.

Adding traces:

```
2023-04-16 20:41:04,906 [ERROR] besm.ProcessP1SequencesThread (7024) Exception processing sequences: <class 'ssl.SSLEOFError'>
2023-04-16 20:41:04,906 [ERROR] besm.ProcessP1SequencesThread (7024) Stack Trace
Traceback (most recent call last):
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\threads.py", line 126, in run
    self.globalConfiguration.scheduler.processP1(p1Sequence)
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\scheduler.py", line 27, in processP1
    self.processor.processSequence(p1Sequence, applyToSchedule)
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\processors.py", line 41, in processSequence
    self.processInformation(self._processorConfig["topics"][obisCode], p1Sequence.getInformationValue(obisCode), p1Sequence.getInformationUnit(obisCode))
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\processors.py", line 213, in processInformation
    self._mqttClient.publish(topic=processLabel, payload=str(processValue))
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 1257, in publish
    rc = self._send_publish(
         ^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 2693, in _send_publish
    return self._packet_queue(PUBLISH, packet, mid, qos, info)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 3016, in _packet_queue
    return self.loop_write()
           ^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 1577, in loop_write
    rc = self._packet_write()
         ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 2464, in _packet_write
    write_length = self._sock_send(
                   ^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 649, in _sock_send
    return self._sock.send(buf)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\ssl.py", line 1210, in send
    return self._sslobj.write(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2393)
2023-04-16 20:41:04,984 [INFO] besm.ProcessP1SequencesThread (7024) Stopped
2023-04-16 20:41:05,431 [WARNING] besm (7280) Waiting for threads to terminate...
2023-04-16 20:41:05,431 [INFO] besm.ReadFromCOMPortThread (7280) Closing Serial Port
2023-04-16 20:41:05,431 [INFO] besm.ReadFromCOMPortThread (8548) Closing Serial Port
2023-04-16 20:41:05,431 [INFO] besm.ParseP1RawDataThread (9468) Stopped
2023-04-16 20:41:05,431 [INFO] besm.ReadFromCOMPortThread (8548) Stopped
2023-04-16 20:41:09,805 [INFO] besm.HealthControllerThread (10196) Stopped
2023-04-16 20:41:09,805 [WARNING] besm (7280) All Threads terminated, relaunching...
2023-04-16 20:41:09,805 [INFO] besm (7280) Closing processors
Traceback (most recent call last):
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\main.py", line 80, in <module>
    globalConfiguration.closeProcessors()
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\configuration.py", line 96, in closeProcessors
    self._processors[processorName].closeProcessor()
  File "C:\Users\Administrateur.AMD\Documents\belgian-smartmeter-p1-to-mqtt\besmreader\processors.py", line 216, in closeProcessor
    self._mqttClient.disconnect()
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 1358, in disconnect
    return self._send_disconnect(reasoncode, properties)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 2853, in _send_disconnect
    return self._packet_queue(command, packet, 0, 0)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 3016, in _packet_queue
    return self.loop_write()
           ^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 1577, in loop_write
    rc = self._packet_write()
         ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 2464, in _packet_write
    write_length = self._sock_send(
                   ^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\paho\mqtt\client.py", line 649, in _sock_send
    return self._sock.send(buf)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\ssl.py", line 1210, in send
    return self._sslobj.write(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLError: [SSL: BAD_LENGTH] bad length (_ssl.c:2393)
```

--- a/besmreader/processors.py
+++ b/besmreader/processors.py
@@ -1,6 +1,7 @@
 from abc import abstractmethod
 from paho.mqtt import client as paho
 from jsonschema import validate as jsvalidate
+from datetime import datetime, timedelta
 import json
 
 from .helper import LoggedClass
@@ -105,6 +106,7 @@ class MQTTP1Processor (P1Processor, LoggedClass):
 
         self.__already_connected_once = False
         self.__assume_connected = False
+        self.__cooldown = None
         pahoClientInit = dict()
 
         pahoClientInit["transport"] = "tcp"
@@ -200,16 +202,19 @@ class MQTTP1Processor (P1Processor, LoggedClass):
 
     def connectMQTT(self) -> None:
         try:
-            if (not self.__already_connected_once):
-                self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)
-                self.__already_connected_once = True
-                super().logger.info('MQTT connected for the first time')
-            else:
-                self._mqttClient.reconnect()
-                super().logger.info('MQTT reconnected')
-            self.__assume_connected = True
+            if ((self.__cooldown is None) or (datetime.now() >= self.__cooldown)):
+                if (not self.__already_connected_once):
+                    self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)
+                    self.__already_connected_once = True
+                    super().logger.info('MQTT connected for the first time')
+                else:
+                    self._mqttClient.reconnect()
+                    super().logger.info('MQTT reconnected')
+                self.__assume_connected = True
+                self.__cooldown = None
         except:
-            super().logger.error('MQTT connect failed')
+            super().logger.error('MQTT (re)connect failed. Cooling down for 2 seconds.')
+            self.__cooldown = datetime.now() + timedelta(seconds = 2)
 
     def processInformation(self, processLabel: str, processValue: str, processUnit: str) -> None:
         try:

--- a/besmreader/processors.py
+++ b/besmreader/processors.py
@@ -205,15 +205,25 @@ class MQTTP1Processor (P1Processor, LoggedClass):
         self.connectMQTT()
 
     def connectMQTT(self) -> None:
-        self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)    
+        try:
+            self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)    
+        except:
+            super().logger.error('MQTT connect failed')
 
     def processInformation(self, processLabel: str, processValue: str, processUnit: str) -> None:
-        if (not self._mqttClient.is_connected):
-            self.connectMQTT()
-        self._mqttClient.publish(topic=processLabel, payload=str(processValue))
+        try:
+            if (not self._mqttClient.is_connected):
+                self.connectMQTT()
+            self._mqttClient.publish(topic=processLabel, payload=str(processValue))
+        except:
+            super().logger.error('MQTT publish failed for %s %s', str(processValue), str(processUnit))
+            super().logger.info('MQTT will try to reconnect on next schedule')
 
     def closeProcessor(self) -> None:
-        self._mqttClient.disconnect()
+        try:
+            self._mqttClient.disconnect()
+        except:
+            super().logger.error('MQTT disconnect failed')
     
     @staticmethod
     def getConfigurationName() -> str:

--- a/besmreader/processors.py
+++ b/besmreader/processors.py
@@ -99,17 +99,12 @@ class MQTTP1Processor (P1Processor, LoggedClass):
         A P1 Port Information processor that sends data to MQTT
     """
 
-    @staticmethod
-    def on_connect(client: paho.Client, userdata, flags, rc) -> None:
-        if (rc==0):
-            super().logger.info("MQTT: Connected successfully")
-        else:
-            super().logger.error("MQTT: Bad connection Returned code rc=", rc)
-
     def __init__(self, processorConfig: dict) -> None:
         P1Processor.__init__(self, processorConfig)
         LoggedClass.__init__(self)
 
+        self.__already_connected_once = False
+        self.__assume_connected = False
         pahoClientInit = dict()
 
         pahoClientInit["transport"] = "tcp"
@@ -130,7 +125,6 @@ class MQTTP1Processor (P1Processor, LoggedClass):
             pahoClientInit["protocol"] = paho.__dict__[self._processorConfig["protocol"]]
 
         self._mqttClient = paho.Client(**pahoClientInit)
-        self._mqttClient.on_connect = MQTTP1Processor.on_connect
         
         #
         # Websockets endpoint and headers configuration
@@ -206,18 +200,26 @@ class MQTTP1Processor (P1Processor, LoggedClass):
 
     def connectMQTT(self) -> None:
         try:
-            self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)    
+            if (not self.__already_connected_once):
+                self._mqttClient.connect(self._processorConfig['broker'], self._processorConfig['port'], 60)
+                self.__already_connected_once = True
+                super().logger.info('MQTT connected for the first time')
+            else:
+                self._mqttClient.reconnect()
+                super().logger.info('MQTT reconnected')
+            self.__assume_connected = True
         except:
             super().logger.error('MQTT connect failed')
 
     def processInformation(self, processLabel: str, processValue: str, processUnit: str) -> None:
         try:
-            if (not self._mqttClient.is_connected):
+            if (not self.__assume_connected):
                 self.connectMQTT()
             self._mqttClient.publish(topic=processLabel, payload=str(processValue))
         except:
             super().logger.error('MQTT publish failed for %s %s', str(processValue), str(processUnit))
             super().logger.info('MQTT will try to reconnect on next schedule')
+            self.__assume_connected = False
 
     def closeProcessor(self) -> None:
         try:

